File: src/model/order.rs
use super::{render_date, render_date_time};
use mysql::prelude::FromRow;
use rocket::time::{Date, PrimitiveDateTime};
use serde::{ser::SerializeMap, Serialize};

pub struct Order {
    id: usize,
    envelope_id: Option<usize>,
    current_location: usize,
    source_location: usize,
    receiver: usize,
    order_type: usize,
    customer: usize,
    date_received: Date,
    date_due: Date,
    date_complete: Option<Date>,
    paid: bool,
    rush: bool,
    picked_up: bool,
    formatted_id: String,
}

pub struct OrderNote {
    id: usize,
    order: usize,
    creator: usize,
    date_time: PrimitiveDateTime,
    note: String,
}

pub struct OrderType {
    id: usize,
    name: String,
}

impl Order {
    pub fn id(&self) -> usize {
        self.id
    }

    pub fn envelope_id(&self) -> Option<usize> {
        self.envelope_id
    }

    pub fn current_location(&self) -> usize {
        self.current_location
    }

    pub fn source_location(&self) -> usize {
        self.source_location
    }

    pub fn receiver(&self) -> usize {
        self.receiver
    }

    pub fn order_type(&self) -> usize {
        self.order_type
    }

    pub fn customer(&self) -> usize {
        self.customer
    }

    pub fn date_received(&self) -> Date {
        self.date_received
    }

    pub fn date_due(&self) -> Date {
        self.date_due
    }

    pub fn date_complete(&self) -> Option<Date> {
        self.date_complete
    }

    pub fn paid(&self) -> bool {
        self.paid
    }

    pub fn rush(&self) -> bool {
        self.rush
    }
}

impl FromRow for Order {
    fn from_row_opt(mut row: mysql::Row) -> Result<Self, mysql::FromRowError>
    where
        Self: Sized,
    {
        let id = match row.take("ID") {
            Some(id) => id,
            None => return Err(mysql::FromRowError(row)),
        };
        let envelope_id = match row.take("EnvelopeID") {
            Some(envelope_id) => envelope_id,
            None => return Err(mysql::FromRowError(row)),
        };
        let current_location = match row.take("CurrentLocation") {
            Some(current_location) => current_location,
            None => return Err(mysql::FromRowError(row)),
        };
        let source_location = match row.take("SourceLocation") {
            Some(source_location) => source_location,
            None => return Err(mysql::FromRowError(row)),
        };
        let receiver = match row.take("Receiver") {
            Some(receiver) => receiver,
            None => return Err(mysql::FromRowError(row)),
        };
        let order_type = match row.take("OrderType") {
            Some(order_type) => order_type,
            None => return Err(mysql::FromRowError(row)),
        };
        let customer = match row.take("Customer") {
            Some(customer) => customer,
            None => return Err(mysql::FromRowError(row)),
        };
        let date_received = match row.take("DateReceived") {
            Some(date_received) => date_received,
            None => return Err(mysql::FromRowError(row)),
        };
        let date_due = match row.take("DateDue") {
            Some(date_due) => date_due,
            None => return Err(mysql::FromRowError(row)),
        };
        let date_complete = match row.take("DateComplete") {
            Some(date_complete) => date_complete,
            None => return Err(mysql::FromRowError(row)),
        };
        let paid = match row.take("Paid") {
            Some(paid) => paid,
            None => return Err(mysql::FromRowError(row)),
        };
        let rush = match row.take("Rush") {
            Some(rush) => rush,
            None => return Err(mysql::FromRowError(row)),
        };
        let picked_up = match row.take("PickedUp") {
            Some(picked_up) => picked_up,
            None => return Err(mysql::FromRowError(row)),
        };
        let formatted_id = match row.take("FormattedID") {
            Some(formatted_id) => formatted_id,
            None => return Err(mysql::FromRowError(row)),
        };

        Ok(Order {
            id,
            envelope_id,
            current_location,
            source_location,
            receiver,
            order_type,
            customer,
            date_received,
            date_due,
            date_complete,
            paid,
            rush,
            picked_up,
            formatted_id,
        })
    }
}

impl Serialize for Order {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(12))?;

        map.serialize_entry("id", &self.id)?;
        map.serialize_entry("envelope_id", &self.envelope_id)?;
        map.serialize_entry("current_location", &self.current_location)?;
        map.serialize_entry("source_location", &self.source_location)?;
        map.serialize_entry("receiver", &self.receiver)?;
        map.serialize_entry("type", &self.order_type)?;
        map.serialize_entry("customer", &self.customer)?;
        map.serialize_entry("date_received", &render_date(self.date_received))?;
        map.serialize_entry("date_due", &render_date(self.date_due))?;
        map.serialize_entry(
            "date_complete",
            &self.date_complete.map(|date| render_date(date)),
        )?;
        map.serialize_entry("paid", &self.paid)?;
        map.serialize_entry("rush", &self.rush)?;
        map.serialize_entry("picked_up", &self.picked_up)?;
        map.serialize_entry("formatted_id", &self.formatted_id)?;

        map.end()
    }
}

impl OrderType {
    pub fn id(&self) -> usize {
        self.id
    }

    pub fn name(&self) -> &str {
        &self.name
    }
}

impl FromRow for OrderType {
    fn from_row_opt(mut row: mysql::Row) -> Result<Self, mysql::FromRowError>
    where
        Self: Sized,
    {
        let id = match row.take("ID") {
            Some(id) => id,
            None => return Err(mysql::FromRowError(row)),
        };

        let name = match row.take("Name") {
            Some(name) => name,
            None => return Err(mysql::FromRowError(row)),
        };

        Ok(OrderType { id, name })
    }
}

impl Serialize for OrderType {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(12))?;

        map.serialize_entry("id", &self.id)?;
        map.serialize_entry("name", &self.name)?;

        map.end()
    }
}

impl OrderNote {
    pub fn id(&self) -> usize {
        self.id
    }

    pub fn order(&self) -> usize {
        self.order
    }

    pub fn creator(&self) -> usize {
        self.creator
    }

    pub fn date_time(&self) -> PrimitiveDateTime {
        self.date_time
    }
}

impl FromRow for OrderNote {
    fn from_row_opt(mut row: mysql::Row) -> Result<Self, mysql::FromRowError>
    where
        Self: Sized,
    {
        let id = match row.take("ID") {
            Some(id) => id,
            None => return Err(mysql::FromRowError(row)),
        };
        let order = match row.take("OrderID") {
            Some(order) => order,
            None => return Err(mysql::FromRowError(row)),
        };
        let creator = match row.take("Creator") {
            Some(creator) => creator,
            None => return Err(mysql::FromRowError(row)),
        };
        let date_time = match row.take("DateTime") {
            Some(date_time) => date_time,
            None => return Err(mysql::FromRowError(row)),
        };
        let note = match row.take("Note") {
            Some(note) => note,
            None => return Err(mysql::FromRowError(row)),
        };

        Ok(OrderNote {
            id,
            order,
            creator,
            date_time,
            note,
        })
    }
}

impl Serialize for OrderNote {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut map = serializer.serialize_map(Some(5))?;

        map.serialize_entry("id", &self.id)?;
        map.serialize_entry("order", &self.order)?;
        map.serialize_entry("creator", &self.creator)?;
        map.serialize_entry("date_time", &render_date_time(self.date_time))?;
        map.serialize_entry("note", &self.note)?;
        map.end()
    }
}

File: js/order.js
let locations;
let current_location;

function on_load() {
    // Load order notes
    get(`/data/employees`, (responseText) => {
        let employees = JSON.parse(responseText);
        get(`/data/order/notes?id=${ID}`, (responseText) => {
            let notes = JSON.parse(responseText);
            display_notes(notes, employees);
        }, error);
    }, error);

    // Get order info
    get(`/data/order?id=${ID}`, (responseText) => {
        let order = JSON.parse(responseText);

        // Get order-type information
        let route;
        switch (order.type) {
            case 1:
                route = `/data/order/film?id=${ID}`;
                break;

            case 2:
                route = `/data/order/framing?id=${ID}`;
                break;

            case 3:
                route = `/data/order/photoshoot?id=${ID}`;
                break;

            default:
                return error();
        }

        get(route, (responseText) => {
            order.type_info = JSON.parse(responseText);

            // Get locations
            get(`/data/locations`, (responseText) => {
                locations = JSON.parse(responseText);

                // Get customer info
                get(`/data/customer?id=${order.customer}`, (responseText) => {
                    let customer = JSON.parse(responseText);

                    // Get receiver employee info
                    get(`/data/employee?id=${order.receiver}`, (responseText) => {
                        let employee = JSON.parse(responseText);
                        display_info(order, customer, employee);
                    }, error);
                }, error);
            }, error);
        }, error);
    }, error);
}

function error() {
    alert("There was an error while loading the order")
}

function display_info(order, customer, employee) {
    // Update basic information
    document.getElementById("order-id").innerText = `${order.formatted_id}`;
    if (order.envelope_id == null) {
        document.getElementById("envelope-id").remove();
        document.getElementById("envelope-id-label").remove();
    } else
        document.getElementById("envelope-id").innerText = `${order.envelope_id}`;

    document.getElementById("customer").innerText = customer.name;
    document.getElementById("date-received").innerText = order.date_received;
    document.getElementById("date-due").innerText = order.date_due;
    document.getElementById("rush").innerText = order.rush ? "Yes" : "No";
    document.getElementById("date-completed").innerHTML = order.date_complete == null ? `Not complete&emsp;<button id="mark-complete" class='inline-button' onclick="mark_complete()">Mark Complete</button>` : order.date_complete;
    document.getElementById("receiver").innerText = employee.name;
    document.getElementById("paid").innerHTML = order.paid ? "Yes" : `No&emsp;<button id="mark-paid" onclick="mark_paid()" class='inline-button'>Mark Paid</button>`;

    for (let location of locations) {
        if (location.id == order.source_location) {
            document.getElementById("source-location").innerText = location.name;
            break;
        }
    }

    // Insert edit button and show current location if not picked up
    //         Otherwise, show "picked up"
    let current_location_obj = document.getElementById("current-location");
    if (order.picked_up) {
        current_location_obj.innerText = "Picked up";
    } else {
        let html;
        for (let location of locations) {
            if (location.id == order.current_location) {
                html = location.name;
                current_location = location.id;
                break;
            }
        }

        html += `&emsp;<button class='inline-button' onclick='begin_edit_location()'>Edit</button><button id="mark-picked-up" class='inline-button' onclick="mark_picked_up()">Mark Picked Up</button>`;
        current_location_obj.innerHTML = html;
    }

    // Update specific information
    switch (order.type) {
        case 1:
            display_film_info(order.type_info);
            break;

        case 2:
            display_framing_info(order.type_info);
            break;

        case 3:
            display_photoshoot_info(order.type_info);
            break;

        default:
            alert("Unknown order type");
    }
}

function display_film_info(type_info) {
    let print_type;
    switch (type_info.prints) {
        case 0:
            print_type = "No";
            break;

        case 1:
            print_type = "Matte";
            break;

        case 2:
            print_type = "Glossy";
            break;

        default:
            error();
            return;
    }

    let label_html = document.getElementById("order-labels").innerHTML;
    let input_html = document.getElementById("order-inputs").innerHTML;

    label_html += `<div>Type:</div>`;
    label_html += `<div>Prints:</div>`;
    label_html += `<div>Digital:</div>`;
    label_html += `<div>Color:</div>`;
    label_html += `<div>Number of Rolls:</div>`;
    label_html += `<div>Exposures:</div>`;

    input_html += `<div>Film</div>`;
    input_html += `<div>${print_type}</div>`;
    input_html += `<div>${type_info.digital ? "Yes" : "No"}</div>`;
    input_html += `<div>${type_info.color ? "Yes" : "No (Black & White)"}</div>`;
    input_html += `<div>${type_info.num_rolls}</div>`;
    input_html += `<div>${type_info.exposures}</div>`;

    document.getElementById("order-labels").innerHTML = label_html;
    document.getElementById("order-inputs").innerHTML = input_html;
}

function display_framing_info(type_info) {
    let label_html = document.getElementById("order-labels").innerHTML;
    let input_html = document.getElementById("order-inputs").innerHTML;

    label_html += `<div>Type:</div>`;
    label_html += `<div>Category:</div>`;
    label_html += `<div>Size:</div>`;

    input_html += `<div>Framing</div>`;
    input_html += `<div>${type_info.category}</div>`;
    input_html += `<div>${type_info.width}x${type_info.height}</div>`;

    document.getElementById("order-labels").innerHTML = label_html;
    document.getElementById("order-inputs").innerHTML = input_html;
}

function display_photoshoot_info(type_info) {
    console.log(type_info);

    let photoshoot_type;
    switch (type_info.type) {
        case 1:
            photoshoot_type = "Family Session";
            break;

        case 2:
            photoshoot_type = "Classic Collection";
            break;

        case 3:
            photoshoot_type = "Location Session";
            break;

        case 4:
            photoshoot_type = "Business Headshot";
            break;

        case 5:
            photoshoot_type = "Standard Graduation Photo";
            break;

        case 6:
            photoshoot_type = "Lifestyle Graduation";
            break;

        default:
            error();
            return;
    }

    let label_html = document.getElementById("order-labels").innerHTML;
    let input_html = document.getElementById("order-inputs").innerHTML;

    label_html += `<div>Type:</div>`;
    label_html += `<div>Photoshoot Type:</div>`;
    label_html += `<div>Scheduled Date and Time:</div>`;

    input_html += `<div>Photoshoot</div>`;
    input_html += `<div>${photoshoot_type}</div>`;
    input_html += `<div>${type_info.date_time}</div>`;

    document.getElementById("order-labels").innerHTML = label_html;
    document.getElementById("order-inputs").innerHTML = input_html;
}

function mark_paid() {
    post(`/data/order/paid?id=${ID}`, () => {
        window.location.reload();
    }, update_error);
}

function mark_complete() {
    post(`/data/order/completed?id=${ID}`, () => {
        window.location.reload();
    }, update_error);
}

function mark_picked_up() {
    post(`/data/order/picked_up?id=${ID}`, () => {
        window.location.reload();
    }, update_error);
}

function update_error() {
    alert("There was an error while updating the order")
}

function begin_edit_location() {
    let html = "<select id='current-location-select'>";

    for (let location of locations)
        html += `<option value="${location.id}"${location.id == current_location ? " selected" : ""}>${location.name}</option>`;

    html += "</select>";

    document.getElementById("current-location").innerHTML = html;

    html = "<button onclick='confirm_location();'>Confirm</button>";
    html += "<button onclick='cancel_location();'>Cancel</button>";

    document.getElementById("buttons").innerHTML = html;
}

function cancel_location() {
    let html;
    for (let location of locations) {
        if (location.id == current_location) {
            html = location.name;
            break;
        }
    }

    html += `&emsp;<button style="display: inline; width: min(150px, 25%); margin: 0;" onclick='begin_edit_location()'>Edit</button>`;
    document.getElementById("current-location").innerHTML = html;
    document.getElementById("buttons").innerHTML = "";
}

function confirm_location() {
    post(`/data/order/change_location?id=${ID}&location=${document.getElementById('current-location-select').value}`, () => {
        window.location.reload();
    }, update_error);
}

function display_notes(notes, employees) {
    let html = "";

    for (let employee of employees)
        html += `<option value="${employee.id}">${employee.name}</option>`;

    document.getElementById("creator").innerHTML = html;

    html = "";

    for (let note of notes) {
        let creator;
        for (let employee of employees) {
            if (employee.id == note.creator) {
                creator = employee.name;
                break;
            }
        }

        html += `<div class='note'>`;
        html += `<div class='note-header'>`;
        html += `<div class='note-creator'>${creator}</div>`;
        html += `<div class='note-date'>${note.date_time}</div>`;
        html += `</div>`;
        html += `<div class='note-content'>${note.note}</div>`
        html += `</div>`;
        html += `<hr />`;
    }

    document.getElementById("notes").innerHTML = html;
}

function begin_create_note() {
    document.getElementById("new-note-button").style.display = "none";
    document.getElementById("create-note").style.display = "block";
}

function cancel_create_note() {
    document.getElementById("note").value = "";
    document.getElementById("creator").selectedIndex = 0;
    document.getElementById("create-note").style.display = "none";
    document.getElementById("new-note-button").style.display = "block";
}

function create_note() {
    // Collect and validate the information
    let creator = document.getElementById("creator").value;

    let note = document.getElementById("note").value;
    if (note == "") {
        document.getElementById("note-error").style.display = "block";
        return;
    } else
        document.getElementById("note-error").style.display = "none";

    // Send the request
    post(`/data/order/create_note`, () => {
        location.reload();
    }, create_error, {
        order: ID,
        creator: Number(creator),
        note: note,
    });
}

function create_error() {
    alert("There was an error while creating the note");
}